File: src/s3_source.py
"""
List and stream G-Core CDN access logs from S3.
Paths follow: prefix/logs/YYYY/MM/DD/HH/mm/ss/{edgename}_{cname}_access.log.gz
"""
import gzip
import sys
from datetime import datetime, timezone
from typing import Iterator, List, Optional

import boto3
from botocore import UNSIGNED
from botocore.config import Config
from botocore.exceptions import ClientError


def verify_bucket_access(
    bucket: str,
    region: str = "eu-west-1",
    profile_name: Optional[str] = None,
    prefix: str = "/5gemerge/logs/",
) -> bool:
    """
    Verify read access to the bucket in the given region.
    Performs head_bucket and a minimal list (1 key) under prefix.
    Returns True if access OK, False otherwise. Prints result to stderr.
    """
    try:
        session = boto3.Session(profile_name=profile_name)
        client = session.client("s3", region_name=region)
        client.head_bucket(Bucket=bucket)
        # Optional: confirm we can list under the logs prefix
        resp = client.list_objects_v2(Bucket=bucket, Prefix=prefix, MaxKeys=1)
        count = resp.get("KeyCount", 0)
        print(
            f"OK: Can access bucket s3://{bucket} (region={region}). "
            f"Prefix '{prefix}' has {'≥1 object' if count else 'no objects yet'}.",
            file=sys.stderr,
        )
        return True
    except ClientError as e:
        code = e.response.get("Error", {}).get("Code", "")
        msg = e.response.get("Error", {}).get("Message", str(e))
        print(
            f"ERROR: Cannot access bucket s3://{bucket} (region={region}): {code} - {msg}",
            file=sys.stderr,
        )
        return False
    except Exception as e:
        print(f"ERROR: {type(e).__name__}: {e}", file=sys.stderr)
        return False


def list_raw_keys(
    bucket: str,
    prefix: str,
    max_keys: int = 10,
    profile_name: Optional[str] = None,
) -> List[tuple[str, datetime, int]]:
    """List up to max_keys objects under prefix (any suffix). For debugging."""
    session = boto3.Session(profile_name=profile_name)
    client = session.client("s3", region_name="eu-west-1")
    paginator = client.get_paginator("list_objects_v2")
    out: List[tuple[str, datetime, int]] = []
    for page in paginator.paginate(Bucket=bucket, Prefix=prefix):
        for obj in page.get("Contents") or []:
            out.append((obj["Key"], obj["LastModified"], obj.get("Size", 0)))
            if len(out) >= max_keys:
                return out
    return out


def list_log_objects(
    bucket: str,
    prefix: str = "/5gemerge/logs/",
    since: Optional[datetime] = None,
    suffix: str = "_access.log.gz",
    profile_name: Optional[str] = None,
    use_unsigned: bool = False,
) -> Iterator[tuple[str, datetime, int]]:
    """
    List S3 objects under bucket/prefix that end with suffix.
    Yields (key, last_modified, size) for each object.
    If since is set, only yield objects last_modified >= since.
    """
    config = Config(signature_version=UNSIGNED) if use_unsigned else None
    session = boto3.Session(profile_name=profile_name)
    client = session.client("s3", config=config, region_name="eu-west-1")
    paginator = client.get_paginator("list_objects_v2")
    for page in paginator.paginate(Bucket=bucket, Prefix=prefix):
        for obj in page.get("Contents") or []:
            key = obj["Key"]
            if not key.endswith(suffix):
                continue
            last_modified = obj["LastModified"]
            # Normalize to comparable: S3 returns aware datetime; since is UTC-aware
            if since:
                lm_utc = last_modified if last_modified.tzinfo else last_modified.replace(tzinfo=timezone.utc)
                if lm_utc < since:
                    continue
            size = obj.get("Size", 0)
            yield key, last_modified, size


def download_and_decompress(
    bucket: str,
    key: str,
    profile_name: Optional[str] = None,
    use_unsigned: bool = False,
) -> bytes:
    """Download an object from S3 and, if gzipped, decompress. Return raw bytes (utf-8 decodable)."""
    config = Config(signature_version=UNSIGNED) if use_unsigned else None
    session = boto3.Session(profile_name=profile_name)
    client = session.client("s3", config=config, region_name="eu-west-1")
    response = client.get_object(Bucket=bucket, Key=key)
    body = response["Body"].read()
    # Some objects are named .gz but are plain text (e.g. under 5gemerge); only decompress real gzip
    if key.endswith(".gz") and len(body) >= 2 and body[:2] == b"\x1f\x8b":
        body = gzip.decompress(body)
    return body


def stream_log_lines(
    bucket: str,
    key: str,
    profile_name: Optional[str] = None,
    use_unsigned: bool = False,
    encoding: str = "utf-8",
    errors: str = "replace",
) -> Iterator[str]:
    """Download a log object from S3, decompress if needed, and yield lines."""
    raw = download_and_decompress(bucket, key, profile_name=profile_name, use_unsigned=use_unsigned)
    text = raw.decode(encoding, errors=errors)
    for line in text.splitlines():
        yield line

File: ts/src/s3-source.ts
/**
 * List and stream G-Core CDN access logs from S3.
 */

import {
  S3Client,
  ListObjectsV2Command,
  GetObjectCommand,
  HeadBucketCommand,
} from '@aws-sdk/client-s3';
import { fromIni } from '@aws-sdk/credential-providers';
import { Readable } from 'stream';
import * as zlib from 'zlib';

const REGION = 'eu-west-1';
const GZIP_MAGIC = new Uint8Array([0x1f, 0x8b]);

export interface S3ObjectInfo {
  key: string;
  lastModified: Date;
  size: number;
}

export async function verifyBucketAccess(
  bucket: string,
  region: string = REGION,
  profile?: string,
  prefix: string = '/5gemerge/logs/'
): Promise<boolean> {
  try {
    const client = createS3Client(region, profile);
    await client.send(new HeadBucketCommand({ Bucket: bucket }));
    const list = await client.send(
      new ListObjectsV2Command({ Bucket: bucket, Prefix: prefix, MaxKeys: 1 })
    );
    const count = list.KeyCount ?? 0;
    console.error(
      `OK: Can access bucket s3://${bucket} (region=${region}). Prefix '${prefix}' has ${count ? '≥1 object' : 'no objects yet'}.`
    );
    return true;
  } catch (e: unknown) {
    const err = e as { name?: string; message?: string; $metadata?: { httpStatusCode?: number } };
    const code = err.name ?? '';
    const msg = err.message ?? String(e);
    console.error(`ERROR: Cannot access bucket s3://${bucket} (region=${region}): ${code} - ${msg}`);
    return false;
  }
}

function createS3Client(region: string, profile?: string): S3Client {
  const opts: { region: string; credentials?: ReturnType<typeof fromIni> } = { region };
  if (profile) {
    opts.credentials = fromIni({ profile });
  }
  return new S3Client(opts);
}

export async function listRawKeys(
  bucket: string,
  prefix: string,
  maxKeys: number = 10,
  profile?: string
): Promise<S3ObjectInfo[]> {
  const client = createS3Client(REGION, profile);
  const out: S3ObjectInfo[] = [];
  let continuationToken: string | undefined;
  do {
    const cmd = new ListObjectsV2Command({
      Bucket: bucket,
      Prefix: prefix,
      MaxKeys: maxKeys - out.length,
      ContinuationToken: continuationToken,
    });
    const resp = await client.send(cmd);
    for (const obj of resp.Contents ?? []) {
      if (obj.Key != null) {
        out.push({
          key: obj.Key,
          lastModified: obj.LastModified ?? new Date(0),
          size: obj.Size ?? 0,
        });
        if (out.length >= maxKeys) return out;
      }
    }
    continuationToken = resp.NextContinuationToken;
  } while (continuationToken);
  return out;
}

export async function* listLogObjects(
  bucket: string,
  prefix: string = '/5gemerge/logs/',
  since?: Date,
  suffix: string = '_access.log.gz',
  profile?: string
): AsyncGenerator<S3ObjectInfo> {
  const client = createS3Client(REGION, profile);
  let continuationToken: string | undefined;
  do {
    const cmd = new ListObjectsV2Command({
      Bucket: bucket,
      Prefix: prefix,
      ContinuationToken: continuationToken,
    });
    const resp = await client.send(cmd);
    for (const obj of resp.Contents ?? []) {
      const key = obj.Key;
      if (key == null || !key.endsWith(suffix)) continue;
      const lastModified = obj.LastModified ?? new Date(0);
      if (since && lastModified < since) continue;
      yield {
        key,
        lastModified,
        size: obj.Size ?? 0,
      };
    }
    continuationToken = resp.NextContinuationToken;
  } while (continuationToken);
}

export async function downloadAndDecompress(
  bucket: string,
  key: string,
  profile?: string
): Promise<Buffer> {
  const client = createS3Client(REGION, profile);
  const resp = await client.send(new GetObjectCommand({ Bucket: bucket, Key: key }));
  const body = resp.Body as Readable;
  if (!body) throw new Error('Empty S3 response body');
  const chunks: Buffer[] = [];
  for await (const chunk of body) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  }
  let buf = Buffer.concat(chunks);
  if (key.endsWith('.gz') && buf.length >= 2 && buf[0] === GZIP_MAGIC[0] && buf[1] === GZIP_MAGIC[1]) {
    buf = zlib.gunzipSync(buf);
  }
  return buf;
}

export async function* streamLogLines(
  bucket: string,
  key: string,
  profile?: string,
  encoding: BufferEncoding = 'utf-8'
): AsyncGenerator<string> {
  const raw = await downloadAndDecompress(bucket, key, profile);
  const text = raw.toString(encoding);
  for (const line of text.split(/\r?\n/)) {
    yield line;
  }
}
